spec/Memory_Consistency.tex

```diff
@@ -2,32 +2,768 @@
-  dynamic security requirements as Java does.
-\end{rationale}
-
-Accessing a synchronization variable (\chpl{sync} or \chpl{single}, \rsec{Synchronization_Variables}) is
-the only way to correctly synchronize between two Chapel tasks.
-Such reads and writes serve as memory fences, preventing
-reordering of reads and writes to traditional variables across the
-synchronization variable's access.  When the same memory location
-is written by one task and read by another, the ordering of the
-read relative to the write is undefined unless there is an intervening
-access to a synchronization variable.
+\begin{openissue}
+Should Chapel provide a memory fence that only complete unordered
+operations started by the current task?
+\end{openissue}
+\begin{openissue}
+Should unordered operations on a particular memory address always wait
+for the address to be computed?
+\end{openissue}
+\begin{openissue}
+Does starting a task or joining with a task necessarily wait for
+unordered operations to complete?
+%% sungeun: commenting out for our own notes.
+%%
+%% (It would appear so from the current
+%% description and the C++11 semantics where thread creation and joining
+%% create a \textit{synchronizes with} relationship).
+%% (but the current version does not explicitly depend on the C++11
+%% semantics, so this is not an issue)
+\end{openissue}
+
+Rather than issuing normal loads and stores to read or write local or
+remote memory, a Chapel program can use \textit{unordered} loads and
+stores when preserving sequential program behavior is not
+important.  The following notation for unordered memory operations
+will be used in this section:
+
+\begin{itemize}
+
+  \item $UL(a)$ indicates an \textit{unordered} \textit{load} from a
+  variable at address $a$. $a$ could point to local or remote memory.
+
+  \item $US(a)$ indicates an \textit{unordered} \textit{store} to a
+  variable at address $a$. Again, $a$ could point to local or remote
+  memory.
+
+\end{itemize}
+
+The \textit{unordered} loads and stores $UL(a)$ and $US(a)$ respect
+fences but not program order. As in Section~\ref{memory_order},
+Unordered loads and stores are ordered with SC atomics.  That is,
+unordered loads and stores for a given task are in total order $<_m$
+respecting the following rules which preserve the order of unordered
+loads and stores relative to SC atomic operations:
+
+\begin{itemize}
+  \item If $UL(a)<_pA_{sc}(b)$ then $UL(a)<_mA_{sc}(b)$
+  \item If $US(a)<_pA_{sc}(b)$ then $US(a)<_mA_{sc}(b)$
+  \item If $A_{sc}(a)<_pUL(b)$ then $A_{sc}(a)<_mUL(b)$
+  \item If $A_{sc}(a)<_pUS(b)$ then $A_{sc}(a)<_mUS(b)$
+\end{itemize}
+
+Unordered loads and stores do not preserve sequential program behavior.
+
+\subsection{Unordered Memory Operations Examples}
+
+Unordered operations should be thought of as happening in a way that
+overlaps with the program task. Unordered operations started in
+different program statements can happen at the same time unless an SC
```

Comment:
"at the same time" can't actually happen for a given memory cell on a real computer. "in any order" would be better